mesonbuild/dependencies/base.py

```diff
@@ -576,37 +576,67 @@ def _set_cargs(self):
         self.compile_args = self._convert_mingw_paths(shlex.split(out))
 
     def _search_libs(self, out, out_raw):
-        link_args = []
-        raw_link_args = []
+        '''
+        @out: PKG_CONFIG_ALLOW_SYSTEM_LIBS=1 pkg-config --libs
+        @out_raw: pkg-config --libs
+
+        We always look for the file ourselves instead of depending on the
+        compiler to find it with -lfoo or foo.lib (if possible) because:
+        1. We want to be able to select static or shared
+        2. We need the full path of the library to calculate RPATH values
+        3. De-dup of libraries is easier when we have absolute paths
+
+        Libraries that are provided by the toolchain or are not found by
+        find_library() will be added with -L -l pairs.
+        '''
         # Library paths should be safe to de-dup
-        libpaths = OrderedSet()
-        raw_libpaths = OrderedSet()
+        #
+        # First, figure out what library paths to use. Originally, we were
+        # doing this as part of the loop, but due to differences in the order
+        # of -L values between pkg-config and pkgconf, we need to do that as
+        # a separate step. See:
+        # https://github.com/mesonbuild/meson/issues/3951
+        # https://github.com/mesonbuild/meson/issues/4023
+        #
+        # Separate system and prefix paths, and ensure that prefix paths are
+        # always searched first.
+        prefix_libpaths = OrderedSet()
+        # We also store this raw_link_args on the object later
+        raw_link_args = self._convert_mingw_paths(shlex.split(out_raw))
+        for arg in raw_link_args:
+            if arg.startswith('-L') and not arg.startswith(('-L-l', '-L-L')):
+                prefix_libpaths.add(arg[2:])
+        system_libpaths = OrderedSet()
+        full_args = self._convert_mingw_paths(shlex.split(out))
+        for arg in full_args:
+            if arg.startswith(('-L-l', '-L-L')):
+                # These are D language arguments, not library paths
+                continue
+            if arg.startswith('-L') and arg[2:] not in prefix_libpaths:
+                system_libpaths.add(arg[2:])
+        # Use this re-ordered path list for library resolution
+        libpaths = list(prefix_libpaths) + list(system_libpaths)
```

Comment:
I think this needlessly duplicates all paths in `prefix_libpaths`? Maybe something with an combined ordered dict?
e.g. combined_paths = a copy of prefix_libpaths before the loop and adding to combined_paths also in the loop that parses `out`? This would in effect move all pathes that are only in out to the end relative to those that are in both.

Reply:
No, it does not duplicate any paths. `prefix_libpaths` and `system_libpaths` are disjoint sets. See the loop right above this line. I'm not sure I understand what you mean.

Reply:
right. I somehow overlooked the `arg[2:] not in prefix_libpaths` part. Sorry for the noise.